### PR TITLE
consume bring your own shim(byos)

### PR DESCRIPTION
**Pull request by wli3**
Bring your own shim

If there are shims packaged by convention in nupkg. Shim Repository will simply copy it to the right location.

The query interface ToolPackageInstance will be in charge of finding the shim folder and filter the right RID. Shim Repository will pick the right file after the folder is located since Shim Repository knows the shim name and it also book keep the files at uninstallation.

During development, due to the wrong adapter level. The mock duplicated too much logic. So, I corrected the abstraction level to lower (only create shim). And replaced the existing mock with a much smaller one without any atomic control and file move, copy logic. At the same time. The chmod, which is a IO action, causes problem during tests. So I added adapter layer to it and put it in Util

### CODE REVIEW

**Comment by wli3** [comment]
@dotnet-bot test RHEL7.2 x64 Release Build
@dotnet-bot test Ubuntu x64 Release Build

**Comment by wli3** [comment]
And tested end to end. I created a package with shim. Use this 
(very rough) https://github.com/dotnet/core/compare/master...wli3:sample-byos?expand=1 to get a package. And made an error during packaged shim creation to ensure that's the same file

File: src/dotnet/CommonLocalizableStrings.resx [hunk 1]
```diff
@@ -640,4 +640,10 @@ setx PATH "%PATH%;{0}"
   <data name="FormatVersionIsMissing" xml:space="preserve">
     <value>Format version is missing. This tool may not be supported in this SDK version. Please contact the author of the tool.</value>
   </data>
-</root>
+  <data name="MoreThanOnePackagedShimAvailable" xml:space="preserve">
+    <value>More than one packaged shim available, they are {0}</value>
```
Comment (peterhuene) [inline src/dotnet/CommonLocalizableStrings.resx]: Perhaps just `More than one packaged shim is available: {0}.`

File: src/dotnet/CommonLocalizableStrings.resx [hunk 2]
```diff
@@ -640,4 +640,10 @@ setx PATH "%PATH%;{0}"
...
+  </data>
+  <data name="FailedToReadNuGetLockFile" xml:space="preserve">
+    <value>Failed to read NuGet LockFile for tool package '{0}': {1}.</value>
```
Comment (peterhuene) [inline src/dotnet/CommonLocalizableStrings.resx]: Nit: generally I leave of periods from messages that contain an exception message, as that's likely to have it's own period.

File: src/dotnet/ShellShim/AppHostShimMaker.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+            var appHostDestinationFilePath = shimPath.Value;
+            var appBinaryFilePath = PathUtility.GetRelativePath(appHostDestinationFilePath, entryPoint.Value);
+
+            EmbedAppNameInHost.EmbedAndReturnModifiedAppHostPath(
+                appHostSourceFilePath: appHostSourcePath,
+                appHostDestinationFilePath: appHostDestinationFilePath,
+                appBinaryFilePath: appBinaryFilePath);
+
+            new FilePermissionSetter().SetUserExecutionPermission(appHostDestinationFilePath);
```
Comment (peterhuene) [inline src/dotnet/ShellShim/AppHostShimMaker.cs]: Should this use a passed in `IFilePermissionSetter `?  Perhaps the same one that `ShellShimRepository` was passed?

File: src/dotnet/ToolPackage/ToolPackageInstance.cs [hunk 4]
```diff
@@ -28,6 +32,58 @@ public IReadOnlyList<CommandSettings> Commands
             }
         }
 
+        public IReadOnlyList<FilePath> PackagedShims
```
Comment (peterhuene) [inline src/dotnet/ToolPackage/ToolPackageInstance.cs]: I think this is too complicated (and expensive) for a property.  I think this should either be a `Lazy` initialized field or a method called `GetPackagedSims` to indicate it is performing work (I favor the first solution since this list shouldn't change).

File: test/Microsoft.DotNet.ToolPackage.Tests/LockFileMatcherTests.cs [hunk 5]
```diff
@@ -1,28 +1,82 @@
 // Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Comment (peterhuene) [inline test/Microsoft.DotNet.ToolPackage.Tests/LockFileMatcherTests.cs]: I'm confused by the changes to this file.  Was this file intended to move somewhere given it's no longer a test class?

File: src/dotnet/CommonLocalizableStrings.resx (hunk 1 above)
Comment (wli3) [reply src/dotnet/CommonLocalizableStrings.resx]: and also @KathleenDollard for using facing string review.

File: src/dotnet/CommonLocalizableStrings.resx (hunk 2 above)
Comment (wli3) [reply src/dotnet/CommonLocalizableStrings.resx]: and also @KathleenDollard for using facing string review.

File: src/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 4 above)
Comment (wli3) [reply src/dotnet/ToolPackage/ToolPackageInstance.cs]: i genuinely confused with the line of Get vs property. In my mind most of the query methods in ToolPackageInstance do disk IO. For me that's the bar. However, do you think we should all convert it to method? Or because of most of them are Lazy so that is fine?

File: test/Microsoft.DotNet.ToolPackage.Tests/LockFileMatcherTests.cs (hunk 5 above)
Comment (wli3) [reply test/Microsoft.DotNet.ToolPackage.Tests/LockFileMatcherTests.cs]: You are right, thanks for catching this. I was only look at the right side didn't realize this is a test file. Something is wrong

Comment (wli3) [reply test/Microsoft.DotNet.ToolPackage.Tests/LockFileMatcherTests.cs]: ah... bad bad merge....

Comment (wli3) [reply test/Microsoft.DotNet.ToolPackage.Tests/LockFileMatcherTests.cs]: that file is fixed

File: src/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 4 above)
Comment (peterhuene) [reply src/dotnet/ToolPackage/ToolPackageInstance.cs]: It's not that it's inherently incorrect, I'm just of the opinion that developers tend to think of properties as "cheap" (return a field, for example) or "expensive at most-once" (lazy-init).  Having a property do work every time can be very surprising for users, especially for enumerable properties:

``` 
foreach (var shim in package.PackagedShims)
{
}

...

foreach (var shim in package.PackagedShims)
{
}
```

As someone looking at that code, I'd expect the two enumerables to be the same.  If I saw a method called `GetPackagedShims` or `EnumeratePackagedShims` instead, I'd expect each invocation to return a different enumeration.

Comment (nguerrera) [reply src/dotnet/ToolPackage/ToolPackageInstance.cs]: "expensive-at-most-once" is the right bar for properties. Anything that does a bunch of computation or IO whenever invoked should be a method.

It looks like a field access, so it should behave like a field access. There's nothing immediately wrong looking with the following. (Other than "style nit: use foreach, but use your imagination to think of a case where indexing is actually needed.)

``` C#
for (int i = 0, i < package.PackagedShims.Count; i++)
{
     var x = package.PackagedShims[i];
}
```

But it calls package.get_PackageShims() twice per loop iteration!

File: src/dotnet/CommonLocalizableStrings.resx (hunk 1 above)
Comment (wli3) [reply src/dotnet/CommonLocalizableStrings.resx]: Changed as suggested

File: src/dotnet/CommonLocalizableStrings.resx (hunk 2 above)
Comment (wli3) [reply src/dotnet/CommonLocalizableStrings.resx]: Removed the last period

File: src/dotnet/ShellShim/AppHostShimMaker.cs (hunk 3 above)
Comment (wli3) [reply src/dotnet/ShellShim/AppHostShimMaker.cs]: Changed as suggested

File: src/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 4 above)
Comment (wli3) [reply src/dotnet/ToolPackage/ToolPackageInstance.cs]: Fixed, use Lazy

**Comment by wli3** [comment]
@dotnet-bot test RHEL6 x64 Debug Build

**Comment by wli3** [comment]
@peterhuene i addressed all the review. It is good?

**Comment by wli3** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build
@dotnet-bot test RHEL7.2 x64 Release Build

**Comment by wli3** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build
